cvmfs/util/file_backed_buffer.h

```diff
@@ -0,0 +1,73 @@
+/**
+ * This file is part of the CernVM File System.
+ */
+
+#ifndef CVMFS_UTIL_FILE_BACKED_BUFFER_H_
+#define CVMFS_UTIL_FILE_BACKED_BUFFER_H_
+
+#include <stdint.h>
+
+#include <climits>
+#include <string>
+
+#include "util/mmap_file.h"
+#include "util/single_copy.h"
+
+/**
+ * FileBackedBuffer is a class for smart data writing and reading.
+ * 
+ * It works in 2 modes: memory mode and file mode. It first starts
+ * working in memory mode, but as soon as the buffer size reaches
+ * a configurable threshold, it writes all data into a temporary file,
+ * freeing up the occupied memory. From that point, all further data
+ * are written directly into the file.
+ * 
+ * Reading is done straight from memory or using mmap, depending on the
+ * current mode of operation.  
+ * 
+ */
+class FileBackedBuffer : SingleCopy {
+ public:
+  explicit FileBackedBuffer(uint64_t in_memory_threshold = ULONG_LONG_MAX,
+                            const std::string &tmp_dir = "/tmp/");
+  explicit FileBackedBuffer(const std::string &file_path);
+  ~FileBackedBuffer();
+  void Write(const void *ptr, uint64_t len);
+  int64_t Read(void** ptr, int64_t len);
```

Comment:
I see. My priority was to make the interface simple and usable and I wanted it to be similar to C file interface (fread, fwrite), because everybody knows that one and it's very easy to use. I see that you want the interface to be descriptive and correspond to the stuff that is actually happening inside. I wanted to go in the opposite direction: conceal the inner stuff and make the public interface as easy as possible.